crates/core/src/operations/load_cdf.rs

```diff
@@ -0,0 +1,471 @@
+                match action {
+                    Action::Cdc(f) => cdc_actions.push(f.clone()),
+                    Action::Metadata(md) => {
+                        log::info!("Metadata: {:?}", &md);
+                        if let Some(Some(key)) = &md.configuration.get("delta.enableChangeDataFeed")
+                        {
+                            if key.to_lowercase() == "false" {
+                                return Err(DeltaTableError::ChangeDataNotRecorded {
+                                    version,
+                                    start,
+                                    end,
+                                });
+                            }
+                        };
+                    }
+                    Action::CommitInfo(ci) => {
+                        ts = ci.timestamp.unwrap_or(0);
+                    }
+                    _ => {}
+                }
+            }
+
+            if !cdc_actions.is_empty() {
+                log::debug!(
+                    "Located {} cdf actions for version: {}",
+                    cdc_actions.len(),
+                    version
+                );
+                change_files.push(CdcDataSpec::new(version, ts, cdc_actions))
+            } else {
+                let add_actions = version_actions
+                    .iter()
+                    .filter_map(|a| match a {
+                        Action::Add(a) if a.data_change => Some(a.clone()),
+                        _ => None,
+                    })
+                    .collect::<Vec<Add>>();
+
+                let remove_actions = version_actions
+                    .iter()
+                    .filter_map(|a| match a {
+                        Action::Remove(r) if r.data_change => Some(r.clone()),
+                        _ => None,
+                    })
+                    .collect::<Vec<Remove>>();
+
+                if !add_actions.is_empty() {
+                    log::debug!(
+                        "Located {} cdf actions for version: {}",
+                        add_actions.len(),
+                        version
+                    );
+                    add_files.push(CdcDataSpec::new(version, ts, add_actions));
+                }
+
+                if !remove_actions.is_empty() {
+                    log::debug!(
+                        "Located {} cdf actions for version: {}",
+                        remove_actions.len(),
+                        version
+                    );
+                    remove_files.push(CdcDataSpec::new(version, ts, remove_actions));
+                }
+            }
+        }
+
+        Ok((change_files, add_files, remove_files))
+    }
+
+    #[inline]
+    fn get_add_action_type() -> Option<ScalarValue> {
+        Some(ScalarValue::Utf8(Some(String::from("insert"))))
+    }
+
+    /// Executes the scan
+    pub async fn build(&self) -> DeltaResult<DeltaCdfScan> {
+        let (cdc, add, _remove) = self.determine_files_to_read().await?;
```

Comment:
Why is nothing done with the remove actions?

Reply:
It's a left over from when I thought I'd have to do coarse grained cdc with this, but to my knowledge without coarse grained removes are not used. I just wanted to leave it for now incase I misunderstood this, but I'll remove it if it holds to not be needed.